kernel/arch/riscv/paging.c

```diff
@@ -47,41 +48,291 @@ static page_table_t *alloc_early_pt(void)
    return &early_pt[early_pt_used++];
 }
 
+static ALWAYS_INLINE page_table_t *
+pdir_get_page_table(pdir_t *pdir, ulong vaddr)
+{
+   page_table_t *pt;
+   page_t e;
+
+   for (int level = RV_PAGE_LEVEL; level > 0; level--) {
+
+      e.raw = pdir->entries[PTE_INDEX(level, vaddr)].raw;
+
+      /* Check PTE valid and is not BIG page */
+      if (!e.present || (e.raw & _PAGE_LEAF))
+         return NULL;
+
+      pt = PA_TO_LIN_VA(e.pfn << PAGE_SHIFT);
+      pdir = (pdir_t *)pt;
+   }
+
+   return pt;
+}
+
 bool handle_potential_cow(void *context)
 {
-   NOT_IMPLEMENTED();
+   regs_t *r = context;
+   ulong vaddr = r->sbadaddr;
+   const void *const page_vaddr = (void *)(vaddr & PAGE_MASK);
+
+   page_table_t *pt = pdir_get_page_table(get_curr_pdir(), vaddr);
+   if (!pt)
+      return false;
+
+   if (!(pt->entries[PTE_INDEX(0, vaddr)].raw & PAGE_COW_ORIG_RW))
+      return false; /* Not a COW page */
+
+   const ulong orig_page_paddr = (ulong)
+      pt->entries[PTE_INDEX(0, vaddr)].pfn << PAGE_SHIFT;
+
+   if (pf_ref_count_get(orig_page_paddr) == 1) {
+
+      /* This page is not shared anymore. No need for copying it. */
+
+#if DEBUG_CHECKS
+      const ulong paddr = (ulong)
+         pt->entries[PTE_INDEX(0, vaddr)].pfn << PAGE_SHIFT;
+
+      ASSERT(paddr != KERNEL_VA_TO_PA(&zero_page));
+#endif
+
+      pt->entries[PTE_INDEX(0, vaddr)].wr = true;
+      pt->entries[PTE_INDEX(0, vaddr)].reserved = 0;
+      invalidate_page_hw(vaddr);
+      return true;
+   }
+
+   // Allocate a new page.
+   void *new_page_vaddr = kmalloc(PAGE_SIZE);
+
+   if (!new_page_vaddr) {
+
+      // Out-of-memory case
+      struct task *curr = get_curr_task();
+
+      if (!curr->running_in_kernel) {
```

Comment:
In Tilck the preferred pattern is:
```
if (error-check) {
   // nagative case handling
   return error;
}

// positive code path
```

In this case, the `!running_in_kernel` condition is the positive path, when you return true, so that should be left outside of the IF. Also, if possible, no `else` case after a return. In summary, the following will look nicer:

```
if (curr->running_in_kernel) {
   panic();
}

send_signal();
return true;
```

Reply:
fixed！